File: coconut2d-x/actions/CCMask.cpp
//
//  CCMask.cpp
//  Coconut2D-X
//
//  Created by Moky on 12-11-28.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#include "CCMask.h"

NS_CC_BEGIN

#pragma mark CCMaskTo

CCMaskTo * CCMaskTo::actionWithDuration(ccTime tDuration, const CCRect & tRect)
{
	CCMaskTo * pAction = new CCMaskTo();
	if (pAction && pAction->initWithDuration(tDuration, tRect))
	{
		pAction->autorelease();
		return pAction;
	}
	CC_SAFE_DELETE(pAction);
	return NULL;
}

bool CCMaskTo::initWithDuration(ccTime tDuration, const CCRect & tRect)
{
	if (!CCActionInterval::initWithDuration(tDuration))
	{
		return false;
	}
	m_tEndRect = CC_RECT_PIXELS_TO_POINTS(tRect);
	return true;
}

CCObject * CCMaskTo::copyWithZone(CCZone * pZone)
{
	CCZone * pNewZone = NULL;
	CCMaskTo * pCopy = NULL;
	if (pZone && pZone->m_pCopyObject)
	{
		// incase of being called at sub class
		pCopy = (CCMaskTo *)(pZone->m_pCopyObject);
	}
	else
	{
		pCopy = new CCMaskTo();
		pZone = pNewZone = new CCZone(pCopy);
	}
	
	CCActionInterval::copyWithZone(pZone);
	
	pCopy->initWithDuration(m_fDuration, m_tEndRect);
	
	CC_SAFE_DELETE(pNewZone);
	return pCopy;
}

void CCMaskTo::startWithTarget(CCNode * pTarget)
{
	CCActionInterval::startWithTarget(pTarget);
	
	CCSprite * pSprite = dynamic_cast<CCSprite *>(m_pTarget);
	if (!pSprite)
	{
		return;
	}
	m_tStartRect = pSprite->getTextureRect();
	
	CCPoint deltaXY = ccpSub(m_tEndRect.origin, m_tStartRect.origin);
	CCPoint deltaWH = ccpSub(ccpFromSize(m_tEndRect.size), ccpFromSize(m_tStartRect.size));
	
	m_tDelta = CCRectMake(deltaXY.x, deltaXY.y, deltaWH.x, deltaWH.y);
}

void CCMaskTo::update(ccTime tTime)
{
	CCSprite * pSprite = dynamic_cast<CCSprite *>(m_pTarget);
	if (!pSprite)
	{
		return;
	}
	
	CCRect tTextureRect = CCRectMake(m_tStartRect.origin.x + m_tDelta.origin.x * tTime,
									 m_tStartRect.origin.y + m_tDelta.origin.y * tTime,
									 m_tStartRect.size.width + m_tDelta.size.width * tTime,
									 m_tStartRect.size.height + m_tDelta.size.height * tTime);
	pSprite->setTextureRect(tTextureRect);
}

#pragma mark CCMaskBy

CCMaskBy * CCMaskBy::actionWithDuration(ccTime tDuration, const CCRect & tRect)
{
	CCMaskBy * pAction = new CCMaskBy();
	if (pAction && pAction->initWithDuration(tDuration, tRect))
	{
		pAction->autorelease();
		return pAction;
	}
	CC_SAFE_DELETE(pAction);
	return NULL;
}

void CCMaskBy::startWithTarget(CCNode * pTarget)
{
	CCMaskTo::startWithTarget(pTarget);
	
	m_tDelta = m_tEndRect;
	m_tDelta.size = CCSizeZero;
}

NS_CC_END

File: spriteforest-ios/action/CCMask.m
//
//  CCMaskTo.m
//  SpriteForest
//
//  Created by Moky on 11-12-26.
//  Copyright 2011 Slanissue.com. All rights reserved.
//

#import "CCMask.h"


@implementation CCMaskTo

+ (id) actionWithDuration:(ccTime)t rect:(CGRect)r
{
	return [[[self alloc] initWithDuration:t rect:r] autorelease];
}

- (id) initWithDuration:(ccTime)t rect:(CGRect)r
{
	if ((self = [super initWithDuration:t]))
	{
		endRect_ = CC_RECT_PIXELS_TO_POINTS(r);
	}
	return self;
}

- (void) startWithTarget:(id)aTarget
{
	[super startWithTarget:aTarget];
	
	startRect_ = [target_ textureRect];
	CGPoint endXY = ccp(endRect_.origin.x, endRect_.origin.y);
	CGPoint endWH = ccp(endRect_.size.width, endRect_.size.height);
	
	CGPoint startXY = ccp(startRect_.origin.x, startRect_.origin.y);
	CGPoint startWH = ccp(startRect_.size.width, startRect_.size.height);
	
	CGPoint deltaXY = ccpSub(endXY, startXY);
	CGPoint deltaWH = ccpSub(endWH, startWH);
	delta_ = CGRectMake(deltaXY.x, deltaXY.y, deltaWH.x, deltaWH.y);
}

- (id) copyWithZone:(NSZone *)zone
{
	CCAction *copy = [[[self class] allocWithZone:zone] initWithDuration:[self duration] rect:endRect_];
	return copy;
}

- (void) update:(ccTime)t 
{
	CGRect newTextureRect = CGRectMake((startRect_.origin.x + delta_.origin.x * t),
									   (startRect_.origin.y + delta_.origin.y * t),
									   (startRect_.size.width + delta_.size.width * t),
									   (startRect_.size.height + delta_.size.height * t));
	[target_ setTextureRect:newTextureRect];
}


@end


@implementation CCMaskBy

+ (id) actionWithDuration:(ccTime)t rect:(CGRect)r
{
	return [[[self alloc] initWithDuration:t rect:r] autorelease];
}

- (void) startWithTarget:(id)aTarget
{
	[super startWithTarget:aTarget];
	
	startRect_ = [target_ textureRect];
	endRect_ = CGRectMake(startRect_.origin.x + endRect_.origin.x,
						 startRect_.origin.y + endRect_.origin.y,
						 endRect_.size.width, endRect_.size.height);
	
	CGPoint endXY = ccp(endRect_.origin.x, endRect_.origin.y);
	CGPoint endWH = ccp(endRect_.size.width, endRect_.size.height);
	
	CGPoint startXY = ccp(startRect_.origin.x, startRect_.origin.y);
	CGPoint startWH = ccp(startRect_.size.width, startRect_.size.height);
	
	CGPoint deltaXY = ccpSub(endXY, startXY);
	CGPoint deltaWH = ccpSub(endWH, startWH);
	delta_ = CGRectMake(deltaXY.x, deltaXY.y, deltaWH.x, deltaWH.y);
}

@end

